ant-protocol/src/messages/merkle_tree.rs

```diff
@@ -0,0 +1,1821 @@
+// Copyright 2025 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use rs_merkle::Hasher;
+use serde::{Deserialize, Serialize};
+use std::time::{SystemTime, UNIX_EPOCH};
+use thiserror::Error;
+use xor_name::XorName;
+
+/// Maximum tree depth (supports 65,536 chunks = 256GB)
+pub const MAX_MERKLE_DEPTH: u8 = 16;
+
+/// Minimum number of leaves (chunks) for a Merkle tree
+pub const MIN_LEAVES: usize = 2;
+
+/// Maximum number of leaves (2^16)
+pub const MAX_LEAVES: usize = 1 << MAX_MERKLE_DEPTH;
+
+/// Maximum age of a Merkle payment (one week in seconds)
+/// Payments older than this are considered expired and nodes will reject chunks
+pub const MERKLE_PAYMENT_EXPIRATION: u64 = 7 * 24 * 60 * 60; // 7 days
+
+/// Errors that can occur when working with Merkle trees
+#[derive(Debug, Error)]
+pub enum MerkleTreeError {
+    #[error("Too few leaves: got {got}, minimum is {MIN_LEAVES}")]
+    TooFewLeaves { got: usize },
+    #[error("Too many leaves: got {got}, maximum is {MAX_LEAVES}")]
+    TooManyLeaves { got: usize },
+    #[error("Invalid leaf index: {index} (tree has {leaf_count} leaves)")]
+    InvalidLeafIndex { index: usize, leaf_count: usize },
+    #[error("Invalid midpoint index: {index} (tree has {midpoint_count} midpoints)")]
+    InvalidMidpointIndex { index: usize, midpoint_count: usize },
+    #[error("Invalid proof")]
+    InvalidProof,
+    #[error("Internal error: {0}")]
+    Internal(String),
+}
+
+pub type Result<T> = std::result::Result<T, MerkleTreeError>;
+
+/// A Merkle tree built from XorNames (chunk addresses)
+///
+/// Used for batch payment system where:
+/// - Build tree from chunk XorNames (min: 2, max: 65,536)
+/// - Tree is automatically padded to next power of 2
+/// - Root is committed on-chain for payment
+/// - Intersections at level depth/2 determine candidate pools
+/// - Individual chunk proofs verify chunks belong to paid batch
+pub struct MerkleTree {
+    /// The underlying rs_merkle tree
+    inner: rs_merkle::MerkleTree<Blake2bHasher>,
+
+    /// Original leaf count (before padding)
+    leaf_count: usize,
+
+    /// Tree depth
+    depth: u8,
+
+    /// The root hash of the tree
+    root: XorName,
+
+    /// Salts for each original leaf
+    /// Used to compute salted hashes: hash(chunk || salt)
+    salts: Vec<[u8; 32]>,
```

Comment:
it has been discussed that random `salts` shall be avoided? 

the `quoting_metrics` can be salts anyway ?

Reply:
For best privacy it's best to keep it random as quoting metrics can be guessed